src/base/cache/cache_entry.hh

```diff
@@ -127,6 +152,8 @@ class CacheEntry : public ReplaceableEntry
      */
     bool valid{false};
 
+    BaseIndexingPolicy *indexingPolicy{nullptr};
```

Comment:
How do you think we should proceed here? I am ok with leaving as is and fixing in the future, but it definitely yells ´fix me!' :)

Reply:
I don't quite follow why we need the indexing policy in the cache entry. Isn't it a property of the cache object? The cache object would have access to the indexing policy which should give you the tag directly instead asking the cache entry for the tag. What am I missing?

Reply:
Hey @pranith, the justification is more or less described in one of the commits:

> Exposing the tag of a cache entry through the associative
> cache APIs makes it hard to generalize the cache for
> structured tags. Ultimately the tag should be a property
> of the cache entry and any tag extraction logic (if needed)
> should reside there. In this we can reuse the associative
> cache for different Entry params, each one bearing a different
> representation of a tag

I agree the indexing policy should be a property of the cache. The issue is at the moment the AssociativeCache is also using it to extract the Tag of an entry. Which shouldn't ideally happen as tagging and indexing are supposed to be "orthogonal" concepts. But anyway the tag is stored in the Entry so my reasoning was that considering the current design the Entry should be responsible for extracting the Tag.

To address more directly your question, if we keep the getTag method within the cache, then we need to sort out an extractTag argument and return types

```
virtual Addr
getTag(const Addr addr) const
{
    return indexingPolicy->extractTag(addr);
}
```

This change came from a previous iteration of the PR, which was not using templates. And templatizing a SimObject (IndexingPolicy) was the real pain I wanted to avoid. Moving the getTag within the Entry solved the problem and afterall it made sense as we store the tag in the entry as I said. Of course the price to pay was moving the IndexingPolicy within the entry which is far from ideal. 

I don't know which is the best solution.. We could for example store a pointer to the AssociativeCache within the Entry... Alternatively an extractTag functor. I am open to suggestions so let me know what you think and if you have abetter solution in mind

Reply:
The updated functor implementation is way better, thanks for fixing it @giactra!